[sanitizer_common] Add experimental flag to tweak dlopen(<main program>)

thurstond opened the pull request:
This introduces an experimental flag 'test_only_replace_dlopen_main_program'.
When enabled, this will replace dlopen(main program,...) with dlopen(NULL,...),
which is the correct way to get a handle to the main program.

This can be useful when ASan is statically linked, since dladdr((void*)pthread_join)
or similar will return the path to the main program.

Note that dlopen(main program,...) never ends well:
- PIE in recent glibc versions (glibc bugzilla 24323), or non-PIE: return an error
- PIE in current GRTE and older glibc: attempt to load the main program
  again, leading to reinitializing ASan and failing to remap the shadow
  memory.

thurstond committed:
[sanitizer_common] Add experimental flag to tweak dlopen(<main program>)

This introduces an experimental flag 'test_only_replace_dlopen_main_program'.
When enabled, this will replace dlopen(<main program,...> with dlopen(NULL,...),
which is the correct way to get a handle to the main program.

This can be useful when ASan is statically linked, since dladdr((void*)pthread_join)
or similar will return the path to the main program.

Note that dlopen(<main program>,...) never ends well:
- PIE in recent glibc versions (glibc bugzilla 24323), or non-PIE: return an error
- PIE in current GRTE and older glibc: attempt to load the main program
  again, leading to reinitializing ASan and failing to remap the shadow
  memory.

thurstond committed:
Check validity of filename string before reading

kstoimenov commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6304,10 +6304,38 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
 
 #if SANITIZER_INTERCEPT_DLOPEN_DLCLOSE
+// Returns 1 if key is a suffix of str, 0 otherwise
+static int internal_strcmp_suffix(const char *key, const char *str) {
```
One confusing part for this implementation is that it has the opposite return value convention of strcmp. Maybe rename to internal_is_suffix? Also key can be renamed to suffix.

thurstond replied:
Good idea, thanks! Done in latest patch set.

kstoimenov commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6304,10 +6304,38 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
+  if (internal_strlen(key) > internal_strlen(str))
+    return 0;
+
+  return !internal_strcmp(str + internal_strlen(str) - internal_strlen(key),
```
Maybe internal_strcmp() == 0 is more clear?

thurstond replied:
Done, thanks!

thurstond committed:
internal_is_suffix fixes per Kirill's suggestion
Added test case per Vitaly's suggestion

thurstond commented:
@vitalybuka Latest patch set added a test case: https://github.com/llvm/llvm-project/pull/71715/files#diff-d1de99adc954c53c720052134ed0f36db3ab9b3349dea742896e97c8187cea0c

kstoimenov reviewed (approved):
LGTM

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6304,10 +6304,38 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
+    COMMON_INTERCEPTOR_READ_STRING(ctx, filename, 0);
+#  if SANITIZER_GLIBC
+    if (common_flags()->test_only_replace_dlopen_main_program &&
+        internal_is_suffix(__progname, filename)) {
```
It's probably batter to use dli_fname from dladdr(NULL) instead of __progname?
And we need to apply readlink to filename, not sure about dli_fname

vitalybuka replied:
ReadBinaryName

thurstond replied:
Per off-line discussion, changed to use dladdr of a symbol that appears in the main binary. (Since 'ISO C++ does not allow 'main' to be used by a program [-Werror,-Wmain]', I ended up using 'dladdr((void *)&SanitizerToolName, &info);').

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6304,10 +6304,38 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
 
 #if SANITIZER_INTERCEPT_DLOPEN_DLCLOSE
+// Returns 1 if 'suffix' matches the end of 'str', 0 otherwise
+static int internal_is_suffix(const char *suffix, const char *str) {
```
I guess we usually use `internal_<some libc function>` 

This is a new function, so we can use Camel style with bool as return.
And usually we search the second argument in the first.

thurstond replied:
Function was removed since the new approach uses exact string matching.

vitalybuka commented on compiler-rt/test/sanitizer_common/TestCases/replace_dlopen_main_program_test.cpp:
```diff
@@ -0,0 +1,29 @@
+// RUN: %clangxx %s -o %t
+// RUN: env %tool_options='test_only_replace_dlopen_main_program=true' %run %t
+// RUN: env %tool_options='test_only_replace_dlopen_main_program=false' not %run %t
+// REQUIRES: glibc
```
This this be in Linux folder then?

vitalybuka replied:
with dladdr/readlink it should work on any posix?

thurstond replied:
I moved it to the Linux folder. 'man dladdr' says it is "nonstandard GNU extensions that are also present on Solaris".

thurstond committed:
Use dladdr to obtain the name of the main program, and then
compare the canonicalized path against the canonicalized dlopen parameter.

Also updated test

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_dl.h:
```diff
@@ -0,0 +1,26 @@
+// Returns the path to the shared object or - in the case of statically linked
+// sanitizers
+// - the main program itself, that contains the sanitizer.
+int dladdr_self_fname(const char **fname);
```
We use the CammelStyle for naming

thurstond replied:
Done

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_dl.cpp:
```diff
@@ -0,0 +1,35 @@
+namespace __sanitizer {
+extern const char *SanitizerToolName;
+
+int dladdr_self_fname(const char **fname) {
```
Can you just  return `const char*` or nullptr?

thurstond replied:
Done

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6304,10 +6305,69 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
 
 #if SANITIZER_INTERCEPT_DLOPEN_DLCLOSE
+// Ordinary internal_readlink does not null-terminate strings.
+static int internal_readlink_with_null(const char *path, char *buf,
```
Naming style

thurstond replied:
Function was removed since we are no longer canonicalizing paths

thurstond commented:
Will refactor the interceptor so that the extra code is a no-op if the flag isn't set

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6304,10 +6305,69 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
+    // We care about dlopen(<main program>,...), which will only happen
+    // for statically linked ASan.
+    const char *self_fname;
+    int ret = dladdr_self_fname(&self_fname);
```
Can you just compare dladdr_self_fname result with filename by pointer?

vitalybuka replied:
Assuming we care about case from dladdr

thurstond replied:
Ended up doing strcmp in case they do an intermediate strcpy

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6304,10 +6305,69 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
+#  if !SANITIZER_DYNAMIC
+    VPrintf(1, "dlopen interceptor: filename: %s\n", filename);
+
+    char *filename_canonical = (char *)InternalAlloc(sizeof(char) * 640);
```
dlopen is fragile, so would ask you to remove heap allocations and logging.
I afraid we can get deadlock or similar

thurstond replied:
Done, removed canonicalization

thurstond committed:
Simplify code, assuming that we don't need to canonicalize paths.

Use camelCase

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6307,7 +6308,36 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
+  if (filename) {
+    COMMON_INTERCEPTOR_READ_STRING(ctx, filename, 0);
+
+#  if !SANITIZER_DYNAMIC
```
`if constexpr (!SANITIZER_DYNAMIC)` may work

thurstond replied:
```
error: use of undeclared identifier 'SANITIZER_DYNAMIC'
    if constexpr (!SANITIZER_DYNAMIC) {
                   ^
1 error generated.
```

(It's usually not defined at all.)

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:
```diff
@@ -6307,7 +6308,36 @@ INTERCEPTOR(int, fclose, __sanitizer_FILE *fp) {
+    if (common_flags()->test_only_replace_dlopen_main_program) {
+      VPrintf(1, "dlopen interceptor: filename: %s\n", filename);
+
+      const char *selfFName = dladdrSelfFName();
```
dladdrSelfFName -> DladdrSelfFName

thurstond replied:
Fixed in the latest commit (https://github.com/llvm/llvm-project/pull/71715/commits/85092c57f47b42e85610c202a3f873a9f6fe8958)

thurstond committed:
Capitalized CamelCase
